crates/bevy_sprite/src/render/mod.rs

```diff
@@ -304,26 +336,180 @@ impl Default for SpriteMeta {
 const QUAD_UVS: [Vec2; 4] = [
+    // Top left
     Vec2::new(0., 1.),
+    // Top right
     Vec2::new(1., 1.),
+    // Bottom right
     Vec2::new(1., 0.),
+    // Bottom left
     Vec2::new(0., 0.),
 ];
 
-#[derive(Component, Eq, PartialEq, Copy, Clone)]
+/// Component defining a batch of sprites for the render world
+#[derive(Component)]
 pub struct SpriteBatch {
+    /// The [`SpriteMeta`] vertex data indices
+    range: Range<u32>,
+    /// The texture handle id
     image_handle_id: HandleId,
+    /// Defines if the `range` targets [`SpriteMeta::vertices`] or [`SpriteMeta::colored_vertices`]
     colored: bool,
+    /// Sort key of the batch
+    z_order: f32,
+}
+
+pub fn prepare_sprites(
+    mut commands: Commands,
+    render_device: Res<RenderDevice>,
+    render_queue: Res<RenderQueue>,
+    mut sprite_meta: ResMut<SpriteMeta>,
+    gpu_images: Res<RenderAssets<Image>>,
+    mut extracted_sprites: ResMut<ExtractedSprites>,
+) {
+    // Impossible handle id value
+    const NULL_HANDLE: HandleId = HandleId::Id(Uuid::nil(), u64::MAX);
+
+    // Clear the vertex buffers
+    sprite_meta.vertices.clear();
+    sprite_meta.colored_vertices.clear();
+
+    // Sort sprites by z for correct transparency and then by handle to improve batching
+    let sort = |a: &ExtractedSprite, b: &ExtractedSprite| match a
+        .transform
+        .translation
+        .z
+        .partial_cmp(&b.transform.translation.z)
+    {
+        Some(Ordering::Equal) | None => a.image_handle_id.cmp(&b.image_handle_id),
+        Some(other) => other,
+    };
+    extracted_sprites.sprites.sort_unstable_by(sort);
+    extracted_sprites.colored_sprites.sort_unstable_by(sort);
+
+    for (colored, sprites) in [
+        (false, &extracted_sprites.sprites),
+        (true, &extracted_sprites.colored_sprites),
+    ] {
+        // Impossible starting values that will be replaced on the first iteration
+        let mut current_batch_handle = NULL_HANDLE;
+        let mut current_image_size = Vec2::ZERO;
+        let mut z_order = 0.0;
+
+        // Vertex buffer indices
+        let mut start = 0;
+        let mut end = 0;
+
+        for extracted_sprite in sprites {
+            if extracted_sprite.image_handle_id != current_batch_handle {
+                if let Some(gpu_image) =
+                    gpu_images.get(&Handle::weak(extracted_sprite.image_handle_id))
+                {
+                    current_image_size = gpu_image.size;
+                    current_batch_handle = extracted_sprite.image_handle_id;
+                    if start != end {
+                        commands.spawn().insert(SpriteBatch {
+                            range: start..end,
+                            image_handle_id: current_batch_handle,
+                            colored,
+                            z_order,
+                        });
+                        start = end;
+                    }
+                } else {
+                    // We skip loading images
```

Comment:
The previous comment of `// Skip this item if the texture is not ready` was better. The point is that if the extracted sprite's texture is not ready, it should not be prepared/queued.